### PR TITLE
Forbid network cidr 100.64.0.0/10

### LINKED ISSUE #169: Prohibit user to create shoot within 100.64.0.0/10 cidr
**How to categorize this issue?**

/area networking
/kind enhancement
/priority normal
/platform alicloud

**What would you like to be added**:
Check service/pod/node network cidrs in shoot don't drop into 100.64.0.0/10 range.
**Why is this needed**:
Alicloud use 100.64.0.0/10 to provide VPC accessed services like DNS, Meta service and so one.

### PR DESCRIPTION
**Pull request by jia-jerry**
**How to categorize this PR?**

/area networking
/kind enhancement
/priority normal
/platform alicloud

**What this PR does / why we need it**:
Avoid user to input wrong cidrs for network settings
**Which issue(s) this PR fixes**:
Fixes #169

**Special notes for your reviewer**:

**Release note**:

```improvement user
Users are forbidden to set cidrs for nodes/pods/services which overlap 100.64.0.0/10 because 100.64.0.0/10 is reverved by AliCloud
```

### CODE REVIEW

**Review by rfranzke — COMMENTED** [review]
What about shoots that already use this CIDR? If this change is introduced they cannot be updated/deleted anymore. 👀

**Comment by jia-jerry** [comment]
> What about shoots that already use this CIDR? If this change is introduced they cannot be updated/deleted anymore. 👀

Good question. So far, no shoots are configured with this wrong cidr. I helped all shoots in ac-canary/ac-live to migrate the old original wrong default network setting(100.X.X.X/X) to correct one. If there is a chance that users specified a wrong one, I made a [tool](https://github.wdf.sap.corp/kubernetes-attic/shoot-network-hacker) to migrate. At least, I can be sure all productions are using correct CIDRs.

**Comment by rfranzke** [comment]
I don't get this argumentation, we cannot look at specific/internal landscapes as the Alicloud extension is public/open-source, and there might be other landscapes that run it.

**Comment by jia-jerry** [comment]
I know what your arguments because it is open sourced. I wantted to be arbitrary that shoots with network wrongly set can't be updated. The good thing is we provide an alarm and force users to take actions to avoid unexpected behaviors.

**Comment by rfranzke** [comment]
Sorry, but I have to insist on that one. If we merge this PR then Alicloud shoots already using these CIDRs cannot be updated/deleted. The only option to merge this PR is to provide a document that explains what users have to do in that case (e.g., migrating the CIDRs with your script, or whatever is necessary to unblock the situation). If you want to provide this information, fine, otherwise, I suggest to adapt the implementation to ignore shoots that already have these CIDRs.

**Comment by jia-jerry** [comment]
OK. I will add some logic to check the network only for creation phase.

**Comment by jia-jerry** [comment]
One thing I don't understand, during update/creation, why should we check network originally? For update, node network is immutable(https://github.com/gardener/gardener/blob/master/pkg/apis/core/validation/shoot.go#L142) For creation, node network can't be nil(https://github.com/gardener/gardener/blob/master/pkg/apis/core/validation/shoot.go#L600).  I would suggest to check shoot network in ali-extention only for creation phase. WDYT?

**Comment by rfranzke** [comment]
@jia-jerry 

> I would suggest to check shoot network in ali-extention only for creation phase and also remove nil check logic. WDYT? 

Once set, all CIDRs in `.spec.networking` (`pods`/`services`/`nodes`) are immutable: https://github.com/gardener/gardener/blob/master/pkg/apis/core/validation/shoot.go#L167-L175
So yes, it should be enough to only check the shoot networks in Alicloud extension when a new shoot is created because Gardener already ensures immutability on update. Good suggestion.

> BTW, which check comes first? Gardener adminsion vs Extension validation hook?

First Gardener static validation, then Gardener's own in-tree admission plugin, then external webhooks.

File: pkg/apis/alicloud/validation/shoot.go
Comment (rfranzke) [inline pkg/apis/alicloud/validation/shoot.go:48]: All these fields (nodes/pods/services) are required for Alicloud. So you should enforce that these fields are not nil. Currently, you only check for intersection if they are not nil, but this would allow end-users to create a shoot without the nodes CIDR, for example.

File: pkg/apis/alicloud/validation/shoot.go
Comment (rfranzke) [inline pkg/apis/alicloud/validation/shoot.go:42]: Don't copy the CIDR multiple times but use the `AliCloudVPCCidr` const (same for other two occurrences).

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/apis/alicloud/validation/shoot.go
Comment (jia-jerry) [reply pkg/apis/alicloud/validation/shoot.go:48]: pods/services cidrs are ensured not nil in Gardener core service: https://github.com/gardener/gardener/blob/master/pkg/utils/validation/cidr/disjoint.go#L37
Regarding node cidrs, you are right. Still need to check nil.

Comment (rfranzke) [reply pkg/apis/alicloud/validation/shoot.go:48]: Yes, but this validation will probably go away soon. Actually, that's a good example why any external code should not rely on what Gardener validates and what not - it is not subject to any API contract or reliable in any other way. This kind of invalidates what we discussed earlier, i.e., also on shoot updates the CIDR immutability should be checked.

Comment (jia-jerry) [reply pkg/apis/alicloud/validation/shoot.go:48]: But from my perspective, I would like to check kubernetes network config (nodes/pods/service) in the core because these fields are exposed in Gardener core API. The extenstion validator only validates those specialties for cloudprovider specific.

Comment (jia-jerry) [reply pkg/apis/alicloud/validation/shoot.go:48]: If such kind of validation will not happen in in-tree admission, shall we provide a generic check in common framwork of extension?

Comment (rfranzke) [reply pkg/apis/alicloud/validation/shoot.go:48]: > because these fields are exposed in Gardener core API.

Yes, those are fields in the core, but it is infrastructure-specific whether they are required or not. For most hyperscalers they are required, but in private clouds or custom infrastructures not necessarily. Hence, the validation should happen here to ensure that ALL the fields needed by Alicloud extension are present.

> If such kind of validation will not happen in in-tree admission, shall we provide a generic check in common framwork of extension?

We could think of something like this in the future, but for now it's straight-forward to simply do it here (without going the de-tour of the extension library, waiting for a new Gardener release, and then call the introduced function).

**Comment by rfranzke** [comment]
What's the status of this PR?

**Comment by jia-jerry** [comment]
> What's the status of this PR?

Recently I'm handling some other urgent issues. Will continue next week. Thanks for your concern

**Comment by rfranzke** [comment]
Hi @minchaow, I think this comment https://github.com/gardener/gardener-extension-provider-alicloud/pull/170#pullrequestreview-519455946 is not yet addressed in your PR.

Please only validate that the CIDR is not used if a new shoot is being created OR if the CIDR is trying to be changed in the update request.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**
